### PR TITLE
[#6703] Stickers UI and contract wedding

### LINKED ISSUE #6703: Add emoji support on Linux and Windows
### User Story

As a <user|developer|...>, I want to <task> so that <goal>.

### Description

*Type*: Bug
*Summary*: on Windows and Linux platforms emojis are grey and they are completely different from what we have on Mac OSx.
At least for now would be nice to have consistency between platforms, ideally as it supposed to be in design (issue https://github.com/status-im/status-react/issues/5611)

#### Expected behavior
mac

#### Actual behavior
**Windows:**
status 2018-11-13 12 55 45
**Ubuntu:**
ubuntu 16 04 2018-11-13 13-56-43

### Reproduction

- Open Status
- Join any public channel
- Send several messages which contain emojis only, text+emoji

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: nighlty 13/11/2018, Windows build https://github.com/status-im/status-react/pull/6465#issuecomment-437633947

[comment]: # (Android or iOS.)
* Operating System: Windows, Linux

### LINKED ISSUE #7453: Stickers UI and contract wedding
Contract: https://ropsten.etherscan.io/address/0x82694e3deabe4d6f4e6c180fe6ad646ab8ef53ae

### PR DESCRIPTION
**Pull request by flexsurfer**
fixes #7453

QA: we have a contract only in ropsten, so you may test this PR in ropsten and buy stickers for STT

https://www.youtube.com/watch?v=L1UMVKOZ2j4

### CODE REVIEW

File: src/status_im/stickers/core.cljs [hunk 1]
```diff
@@ -17,24 +28,102 @@
...
+(re-frame/reg-fx
+ :stickers/pack-data-fx
+ (fn [[web3 network id open?]]
+   (when-let [contract (get ethereum.stickers/contracts (ethereum/network->chain-keyword network))]
+     (ethereum.stickers/pack-data web3 contract id (pack-data-callback id open?)))))
+
+(re-frame/reg-fx
+ :stickers/set-pending-timout-fx
+ (fn []
+   (js/setTimeout #(re-frame/dispatch [:stickers/pending-timout]) 10000)))
```
Comment (jeluard) [inline src/status_im/stickers/core.cljs:132]: Don't we have a generic effect for this?

**Review by jeluard — APPROVED** [review]

File: src/status_im/stickers/core.cljs (hunk 1 above)
Comment (flexsurfer) [reply src/status_im/stickers/core.cljs:132]: haven't seen, i'll take a look thanks

**Comment by flexsurfer** [comment]
@dmitryn @goranjovic guys could you take a look please

File: src/status_im/events.cljs [hunk 2]
```diff
@@ -1675,14 +1673,22 @@
...
+(handlers/register-handler-fx
+ :stickers/load-pack
+ (fn [_ [_ proto-code hash id price open?]]
+   {:http-get {:url                   (str (if (= constants/swarm-proto-code proto-code)
+                                             "https://swarm-gateways.net/bzz:/"
+                                             "https://ipfs.infura.io/ipfs/")
+                                           hash)
+               :success-event-creator (fn [o]
+                                        [:stickers/load-sticker-pack-success o id price open?])
+               :failure-event-creator (fn [o] nil)}}))
```
Comment (cammellos) [inline src/status_im/events.cljs]: `(constantly nil)` is probably more idiomatic

File: src/status_im/events.cljs [hunk 3]
```diff
@@ -1691,5 +1697,36 @@
...
+(handlers/register-handler-fx
+ :stickers/get-owned-packs
+ (fn [{{:keys [web3 network] :as db} :db} _]
+   (let [address (ethereum/normalized-address (get-in db [:account/account :address]))]
+     {:stickers/owned-packs-fx [web3 network address]})))
+
+(handlers/register-handler-fx
+ :stickers/pack-owned
+ (fn [{db :db} [_ id]]
+   {:db (update db :stickers/packs-owned conj id)}))
```
Comment (cammellos) [inline src/status_im/events.cljs]: We generally keep the implementation in the the `core` file

**Review by cammellos — APPROVED** [review]
Overall looks ok to me,  I think we should add some tests, at least for the core functions.

File: src/status_im/stickers/core.cljs (hunk 1 above)
Comment (yenda) [reply src/status_im/stickers/core.cljs:132]: we do in utils there is already a set-timeout

File: src/status_im/utils/ethereum/stickers.cljs [hunk 4]
```diff
@@ -0,0 +1,30 @@
+(ns status-im.utils.ethereum.stickers
+  (:require [status-im.utils.ethereum.core :as ethereum]
+            [status-im.utils.ethereum.abi-spec :as abi-spec]))
+
+(def contracts
```
Comment (yenda) [inline src/status_im/utils/ethereum/stickers.cljs:5]: can you add docstrings to these functions?

**Comment by asemiankevich** [comment]
@flexsurfer seems you are reffering to wrong issue. Please correct so the bug is not closed by accident

**Comment by asemiankevich** [comment]
also @flexsurfer 

- why buying stickers require 2 transactions? Nevermind it was sign message to dapp. 
- why second transaction does not fetch gas limit?

Steps i did: open status -> 1x1 chat -> stickers -> get sticker pack for 15 SNT -> sign transaction -> transaction appears with empty gas limit

image

**Comment by flexsurfer** [comment]
@asemiankevich thanks! fixed issue number. We have 2 transactions because in the first one you approve SNT amount in the SNT contract, and second, buy a stickers pack in stickers registry contract, but @3esmit is working on a solution to have only one, we'll do it later in a separate issue.
For some reason in ropsten sometimes gas limit estimation doesn't work properly, maybe @3esmit could answer why

**Comment by asemiankevich** [comment]
> @asemiankevich thanks! fixed issue number. We have 2 transactions because in the first one you approve SNT amount in the SNT contract, and second, buy a stickers pack in stickers registry contract, but @3esmit is working on a solution to have only one, we'll do it later in a separate issue.
> For some reason in ropsten sometimes gas limit estimation doesn't work properly, maybe @3esmit could answer why

so should i wait for network? As a user i can't buy stickers now (

**Comment by flexsurfer** [comment]
@asemiankevich you can, just set something like 700000, in mainnet it should work fine

**Comment by asemiankevich** [comment]
eh true, so i can buy stickers. @flexsurfer @errorists what you guys think if we add a pop-up (like a tip) saying user that he does not have enough SNT to buy stickers if balance is not sufficient? 

So if i have 0 STT: i can see the button is grayed out however if i tap it -> i am prompted to perform transactions, the first one fails due to `Error encountered during contract execution` and second is confirmed. This seems to be wrong. Please advise

image

**Comment by flexsurfer** [comment]
@asemiankevich it's wrong for sure, this button shouldn't be pressable, i'll fix that

**Comment by flexsurfer** [comment]
@asemiankevich 1-  are you sure you have 0 STT? if you have 0 it should be "buy with snt", 2- why for 15snt pack button without a label?

**Comment by asemiankevich** [comment]
> @asemiankevich 1- i you sure you have 0 STT? if you have 0 it should be "buy with snt", 2- why for 15snt pack button without a label?

- i am using account without STT, only ETH
- label become like that when i tapped it and completed transactions

**Review by goranjovic — APPROVED** [review]

**Comment by flexsurfer** [comment]
@asemiankevich should be fixed, grey buttons should be disabled, for some reason after rebase they got broken, sorry

**Comment by asemiankevich** [comment]
@flexsurfer some builds are failed (ios and e2e both android and ios)

**Comment by flexsurfer** [comment]
@asemiankevich maybe because of rebasing again, can we try to re-run these builds?

**Comment by asemiankevich** [comment]
> @asemiankevich maybe because of rebasing again, can we try to re-run these builds?

yes, i restarted.

**Comment by asemiankevich** [comment]
@flexsurfer second transaction fails and sticker pack is shown with endless spinner
image
 can we handle this too somehow?

**Comment by flexsurfer** [comment]
@asemiankevich I believe we can handle this, but better in a separate issue , thanks

**Comment by asemiankevich** [comment]
> @asemiankevich I believe we can handle this, but better in a separate issue , thanks

alright lets merge it then

**Pull request merged**